crates/dbsp/src/circuit/checkpointer.rs

```diff
@@ -306,3 +309,123 @@ impl Checkpointer {
         }
     }
 }
+
+/// Trait for types that can be check-pointed and restored.
+///
+/// This is to be used for small state within circuit operators
+/// (e.g., everything that's not stored within a batch which
+/// is already stored in files).
+pub trait Checkpoint {
+    fn checkpoint(&self) -> Vec<u8>;
+    fn restore(&mut self, data: &[u8]);
+}
+
+impl Checkpoint for isize {
+    fn checkpoint(&self) -> Vec<u8> {
+        todo!()
+    }
+
+    fn restore(&mut self, _data: &[u8]) {
+        todo!()
+    }
+}
+
+impl Checkpoint for usize {
+    fn checkpoint(&self) -> Vec<u8> {
+        todo!()
+    }
+
+    fn restore(&mut self, _data: &[u8]) {
+        todo!()
+    }
+}
+
+impl Checkpoint for i32 {
+    fn checkpoint(&self) -> Vec<u8> {
+        todo!()
+    }
+
+    fn restore(&mut self, _data: &[u8]) {
+        todo!()
+    }
+}
+
+impl<N> Checkpoint for Box<N>
+where
+    N: Checkpoint + ?Sized,
+{
+    fn checkpoint(&self) -> Vec<u8> {
+        self.as_ref().checkpoint()
+    }
+
+    fn restore(&mut self, data: &[u8]) {
+        self.as_mut().restore(data);
+    }
+}
+
+impl Checkpoint for dyn dynamic::data::Data + 'static {
+    fn checkpoint(&self) -> Vec<u8> {
+        let mut s = Serializer::default();
+        let _r = self.serialize(&mut s).unwrap();
+        let fbuf = s.into_serializer().into_inner();
+        fbuf.into_vec()
+    }
+
+    fn restore(&mut self, data: &[u8]) {
+        unsafe { self.deserialize_from_bytes(data, 0) };
+    }
+}
+
+impl Checkpoint for dyn DataTyped<Type = u64> + 'static {
+    fn checkpoint(&self) -> Vec<u8> {
+        todo!()
+    }
+
+    fn restore(&mut self, _data: &[u8]) {
+        todo!()
+    }
+}
+
+impl<K, V, R, B> Checkpoint for TypedBatch<K, V, R, B>
```

Comment:
This and the following impls contradict the comment that `Checkpoint` is for small state only.